### ISSUE TITLE
New rule proposal: Don't use HasValue property on values of nullable types

### ISSUE THREAD

**Issue by alxyr** [DotNetAnalyzers/StyleCopAnalyzers#1455]
**Category**: Readability
**Rule**: Do not use HasValue property on values of nullable types.
**Description**: Compare nullable values to null instead of using HasValue property.
**Rationale**: Nullable values support reference type semantics and there is no reason to use a special property (HasValue) to test whether a variable is initialized. Comparison to null should be used instead for both value-typed and reference-typed variables.
**See also**: SA1125:UseShorthandForNullableTypes

The warning should be generated in the code like this:
```C#
if (name == null)
	DoSomething();
if (!amount.HasValue) // warning here (compare to null instead)
	DoSomethingElse();
```
Correct code should look like this:
```C#
if (name == null)
	DoSomething();
if (amount == null)
	DoSomethingElse();
```

**Comment by sharwell** [comment]
I'm currently neutral on this. It's an interesting proposal for sure.

**Comment by otac0n** [comment]
This was in our style guide at a previous employer , so I'm :+1:

**Comment by nbarbettini** [comment]
Leaning toward :+1:. I'm used to `!amount.HasValue`, out of habit, but I like the reasoning here.

**Comment by nbarbettini** [comment]
@vweijsters Any opinion on this?

**Comment by otac0n** [comment]
Which one is closer to the way the actual implementation behaves?  Does the JIT emit a null check?

**Comment by alxyr** [comment]
The generated IL is exactly the same, whether the code is written as `!amount.HasValue` or `amount == null`, so the JIT will not see a difference.

**Comment by vweijsters** [comment]
:+1: for me, but with a very minor preference.
